src/main/java/mekanism/common/integration/computer/TableType.java

```diff
@@ -0,0 +1,73 @@
+import mekanism.common.util.MekCodecs;
+
+/**
+ * Defines the format of a "table" (Map) format of a Mekanism data structure
+ */
+public record TableType(String description, String humanName, Map<String, FieldType> fields, Class<?> extendz) {
+
+    public static Codec<TableType> CODEC = RecordCodecBuilder.create(instance ->
+          instance.group(
+                Codec.STRING.fieldOf("description").forGetter(TableType::description),
+                Codec.STRING.fieldOf("humanName").forGetter(TableType::humanName),
+                Codec.unboundedMap(Codec.STRING, FieldType.CODEC).optionalFieldOf("fields", Collections.emptyMap()).forGetter(TableType::fields),
+                MekCodecs.CLASS_TO_STRING_CODEC.optionalFieldOf("extends", null).forGetter(TableType::extendz)
+          ).apply(instance, TableType::new)
+    );
+    public static Codec<Map<Class<?>, TableType>> TABLE_MAP_CODEC = Codec.unboundedMap(MekCodecs.CLASS_TO_STRING_CODEC, CODEC);
+
+    public static Builder builder(Class<?> clazz, String description) {
+        return new Builder(clazz, description);
+    }
+
+    public record FieldType(String description, Class<?> javaType, String type, Class<?>[] javaExtra){
+        public static final Codec<FieldType> CODEC = RecordCodecBuilder.create(instance ->
+              instance.group(
+                    Codec.STRING.fieldOf("description").forGetter(FieldType::description),
+                    MekCodecs.CLASS_TO_STRING_CODEC.fieldOf("javaType").forGetter(FieldType::javaType),
+                    Codec.STRING.fieldOf("type").forGetter(FieldType::type),
+                    MekCodecs.optionalClassArrayCodec("javaExtra").forGetter(FieldType::javaExtra)
+              ).apply(instance, FieldType::new)
+        );
+    }
+
+    public static class Builder {
+
+        private final Class<?> clazz;
+        private final String description;
+        private final String humanName;
+        private final Map<String, FieldType> fields = new LinkedHashMap<>();
+        private Class<?> extendz = null;
+
+        private Builder(Class<?> clazz, String description) {
+            this.clazz = clazz;
+            this.description = description;
+            this.humanName = MethodHelpData.getHumanType(clazz, ComputerMethodFactory.NO_CLASSES);
+        }
+
+        public Builder extendz(Class<?> c) {
+            this.extendz = c;
+            return this;
+        }
+
+        public Builder addField(String name, Class<?> javaType, String description, Class<?>... javaExtra) {
+            if (javaExtra == null) {
+                javaExtra = ComputerMethodFactory.NO_CLASSES;
+            }
+            this.fields.put(name, new FieldType(description, javaType, MethodHelpData.getHumanType(javaType, javaExtra), javaExtra));
+            return this;
+        }
+
+        public TableType build(Map<Class<?>, TableType> destination) {
+            TableType tableType = new TableType(description, humanName, fields, extendz);
```

Comment:
Do we want to ensure a copy on `fields` here? Given otherwise technically if we ended up re-using a builder and add a field after having built it once then it will also be added to the already built variant. Otherwise we can just add docs to the methods that after `build` is called none of the other methods should be used.